spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,201 @@
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import scala.reflect.ClassTag
+
+/** Create Cassandra data source relation based on [[ScanType]]*/
+case class CassandraSourceRelation(table: String, keyspace: String, scanType: ScanType, cluster: Option[String],
+                                   userSpecifiedSchema: Option[StructType], sqlContext: SQLContext) {
+  def apply(table: String, keyspace: String, scanType: ScanType, cluster: Option[String],
+            userSpecifiedSchema: Option[StructType], sqlContext: SQLContext) = scanType.makeRelation(
+    table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Base table scan relation*/
+case object BaseScanType extends ScanType  {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new TableScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns table scan relation*/
+case object PrunedScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new PrunedScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and filtered table scan relation*/
+case object PrunedFilteredScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new PrunedFilteredScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and Catalyst filtered table scan relation*/
+case object CatalystScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new CatalystScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Table scan type */
+sealed trait ScanType {
+  def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                   userSpecifiedSchema: Option[StructType], sqlContext: SQLContext) : BaseRelationImpl
+}
+
+/** Base relation implements [[BaseRelation]] and [[InsertableRelation]] */
+private[cassandra] class BaseRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                          userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelation with InsertableRelation with Serializable with Logging {
+
+  override def sqlContext: SQLContext = sqlcontext
```

Comment:
Use `override val` in constructor instead of creating a method. It will override `def` in the parent trait.